### ISSUE TITLE
Slow training for tf.keras.layers.Embedding with the new optimizers

### ISSUE THREAD

**Issue by vysarge** [keras-team/keras#18279]
**System information**.
- Have I written custom code (as opposed to using a stock example script provided in Keras): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 20.04.6 LTS
- TensorFlow installed from (source or binary): binary, presumably
- TensorFlow version (use command below): v2.12.0-rc1-12-g0db597d0d75
- Python version: 3.10.12 (main, Jun  7 2023, 12:45:35) [GCC 9.4.0]
- Bazel version (if compiling from source): n/a
- GPU model and memory: NVIDIA Tesla T4 16GB
- Exact command to reproduce: [Code in colab here.](https://colab.research.google.com/drive/1qx5JNlgjIpcdshdAQCpRT3Smxq_etfWm?usp=sharing)

**Describe the problem**.

I'm seeing a significant performance regression when testing training with large embedding tables (tf.keras.layers.Embedding) with some of the new optimizers compared to the legacy optimizers.  Times are logged from the last cell of the colab notebook; for larger tables this can be a severalfold slowdown.  In this tiny test I see the issue with SGD, Adadelta, Adagrad, Adamax, and Ftrl.  Relative to legacy, the new Nadam optimizer performs similarly and the new Adam and RMSProp optimizers speed up.

As a side-note, I'm not certain if this should be filed as a separate issue (and perhaps to XLA directly) but I am also seeing slowdowns when passing `jit_compile=True` to `model.compile` with some of the new optimizers.  In the colab notebook, altering one line:
`model.compile(optimizer=o, loss='binary_crossentropy')`
to
`model.compile(optimizer=o, loss='binary_crossentropy', jit_compile=True)`
in the final cell displays a further slowdown for SGD, Adagrad, and FTRL optimizers, for example.

**Describe the current behavior**.

Rough performance figures summarized from this test (all ms/epoch for the same type of random data):
* SGD: 97.2 legacy vs 605.1 new
* Adadelta: 69.6 legacy vs 842.1 new
* Adagrad: 66.9 legacy vs 88.2 new
* Adamax: 72.7 legacy vs 898.6 new
* FTRL: 71.5 legacy vs 983.2 new

**Standalone code to reproduce the issue**.

[Code in colab here.](https://colab.research.google.com/drive/1qx5JNlgjIpcdshdAQCpRT3Smxq_etfWm?usp=sharing)

**Comment by tilakrayal** [comment]
@vysarge,
I tried to execute the mentioned code multiple iterations with the same tensorflow version and observed that training time was gradually decreased. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/982736e51640e173ef2df39503561f9a/2023_07_06_keras_optimizer_speed_test.ipynb).

The slow training might be happening if there are multiple colabs are running at the same time and connection might be the reason in some cases. I request please try with another virtual environment which might yield good results.  Thank you!

**Comment by vysarge** [comment]
Thanks for taking a look!  Apologies for the delay, I've recently returned from vacation.

I ran the test again using a recent container (`docker.io/tensorflow/tensorflow:latest-gpu@sha256:b4676741c491`) on a machine with no other work being done.  This test used a A100-40GB GPU instead of a T4, so the times differ, but I still see performance regression for those cases:

* SGD: 84.2 legacy vs 172.5 new
* Adadelta: 86.5 legacy vs 223.5 new
* Adagrad: 87.5 legacy vs 107.6 new
* Adamax: 93.2 legacy vs 237.3 new
* FTRL: 93.6 legacy vs 232.2 new

Tensorflow version here was `v2.13.0-rc2-7-g1cb1a030a62` and Python version was `3.8.10`.

> I tried to execute the mentioned code multiple iterations with the same tensorflow version and observed that training time was gradually decreased.

@tilakrayal, I ran this test for a while longer using SGD and didn't see a decrease in training time; how many epochs did you run?  At a glance the times look consistent to me in your colab.

**Comment by tilakrayal** [comment]
@vysarge,
I haven't changed any of the code which was mentioned earlier. The number of epochs we are trying are epochs = 20 which was similar. As mentioned above I tried to execute the mentioned code multiple iterations with the same tensorflow version and observed that training time was gradually decreased. I suggest you try with the other GPUs as well for the difference. Thank you!

**Comment by vysarge** [comment]
I'm sorry, I'm not sure I'm following correctly.  In the gist you shared, it appears to me that most times are similar to the ones in my colab and I don't see a decline in the training time over those 20 epochs.  What am I missing?  Was the result with a decreased training time obtained from running in colab, or elsewhere?

By multiple iterations, am I understanding correctly that you mean running the model.fit several times?  I experimented with repeatedly executing the last cell of the colab (model compile & fit) and also with repeatedly running model.fit on the same model, both on a T4 GPU as before.  After 10-15 iterations this way I didn't see a trend in training time using SGD optimizer for either; the epoch time seems to stabilize around 65ms for the legacy optimizer and 550ms for the new optimizer.  I do see minor variation between runs, but it looks random to me.

Or, if this test was done on a different setup, are you able to share what type of GPU was used?  I can also try another GPU type on my side, but I'd like to make sure I can replicate your result.

Thanks!
